### PR TITLE
Path Sequences

### PR DESCRIPTION
**Pull request by ettorealdrovandi**
It took a long time, but this is an almost word-for-word port of the path sequences library in https://github.com/HoTT/HoTT-Agda.

I have tried to limit the overuse of wild unicode characters for function names.

### CODE REVIEW

File: source/PathSequences/Ap.lagda [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+and then as a lift
+
+\begin{code}
+
+  ≡-＝-nat₂ : {x x' : X} {y y' : Y}
+            → (s : x ≡ x') (t : y ≡ y')
+            → ap₂ f [ s ↓] [ t ↓] ＝ [ ap₂-seq s t ↓]
+  ≡-＝-nat₂ [] [] = refl
+  ≡-＝-nat₂ {x} [] (p ◃∙ t) =
+     ap₂ f refl (p ∙ [ t ↓])              ＝⟨ ap₂-refl-left f _ ⟩
```
Comment (martinescardo) [inline source/PathSequences/Ap.lagda:142]: Would you mind using indentation of 1 space?

File: source/PathSequences/Ap.lagda [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+     ap (f x) (p ∙ [ t ↓])                ＝⟨ ap-∙ (f x) p ([ t ↓]) ⟩
+     ap (f x) p ∙ (ap (f x) [ t ↓])       ＝⟨ ap (λ v → (ap (f x) p) ∙ v) (ap-seq-∙-= (f x) t) ⟩
+     ap (f x) p ∙ [ ap-seq (f x) t ↓] ∎
+  ≡-＝-nat₂ {y = y} (p ◃∙ s) [] =  ap₂ f (p ∙ [ s ↓]) refl  ＝⟨ ap₂-refl-right f _ ⟩
```
Comment (martinescardo) [inline source/PathSequences/Ap.lagda:146]: Please indent as you did in line 142 to avoid long lines.

File: source/PathSequences/Ap.lagda [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+                                   ap f₂ (p ∙ [ s ↓])       ＝⟨ ap-∙ f₂ p ([ s ↓]) ⟩
+                                   ap f₂ p ∙ ap f₂ [ s ↓]   ＝⟨ ap (λ v → ap f₂ p ∙ v) (ap-seq-∙-= f₂ s) ⟩
+                                   ap f₂ p ∙ [ (ap-seq f₂ s) ↓] ∎
+       where
+         f₂ = λ v → f v y
+  ≡-＝-nat₂ (p ◃∙ s) (q ◃∙ t) = ap₂ f (p ∙ [ s ↓]) (q ∙ [ t ↓]) ＝⟨ ap₂-∙ f p ([ s ↓]) q ([ t ↓]) ⟩
```
Comment (martinescardo) [inline source/PathSequences/Ap.lagda:152]: Same indentation request.

File: source/PathSequences/Concat.lagda [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+_∙▹_ : {X : 𝓤 ̇ } {x y z : X}
+     → x ≡ y → y ＝ z → x ≡ z
+s ∙▹ p = s ∙≡ (p ◃∎)
+
+≡-to-＝-hom : {X : 𝓤 ̇ } {x y z : X}
+            → (s : x ≡ y) (t : y ≡ z)
+            → ([ s ↓]) ∙ ([ t ↓]) ＝ [ (s ∙≡ t) ↓]
+≡-to-＝-hom [] t = refl-left-neutral
+≡-to-＝-hom (p ◃∙ s) [] =
+              [ (p ◃∙ s) ↓] ∙ [ [] ↓]  ＝⟨ refl-right-neutral [ (p ◃∙ s) ↓] ⁻¹ ⟩
```
Comment (martinescardo) [inline source/PathSequences/Concat.lagda:81]: Indentation of 1 space please.

File: source/PathSequences/Concat.lagda [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+              [ (p ◃∙ s) ↓]            ＝⟨ ap (λ v → [ v ↓]) ([]-∙≡-right-neutral (p ◃∙ s)) ⁻¹ ⟩
+              [ (p ◃∙ s ∙≡ []) ↓]       ∎
+≡-to-＝-hom (p ◃∙ s) (q ◃∙ t) =
+              [ (p ◃∙ s) ↓] ∙ [ (q ◃∙ t) ↓]  ＝⟨ refl ⟩
```
Comment (martinescardo) [inline source/PathSequences/Concat.lagda:85]: Same.

File: source/PathSequences/Inversion.lagda [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+seq⁻¹-∙▹ : {x x' x'' : X} (s : x ≡ x') (p : x' ＝ x'')
+         → seq⁻¹ (s ∙▹ p) ＝ (p ⁻¹) ◃∙ seq⁻¹ s
+seq⁻¹-∙▹ [] p = refl
+seq⁻¹-∙▹ (p₁ ◃∙ s) p = ap (_∙▹ (p₁ ⁻¹)) (seq⁻¹-∙▹ s p)  
+
+seq⁻¹-involutive : {x x' : X} (s : x ≡ x')
+                 → seq⁻¹ (seq⁻¹ s) ＝ s
+seq⁻¹-involutive [] = refl
+seq⁻¹-involutive (p ◃∙ s) = seq⁻¹ (seq⁻¹ s ∙▹ (p ⁻¹))      ＝⟨ seq⁻¹-∙▹ (seq⁻¹ s) (p ⁻¹) ⟩
```
Comment (martinescardo) [inline source/PathSequences/Inversion.lagda:37]: Break line after `=` and indent by 1 please.

File: source/PathSequences/Inversion.lagda [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+          → s ＝ₛ t
+          → seq⁻¹ s ＝ₛ seq⁻¹ t
+seq⁻¹-＝ₛ {s = s} {t} e = seq⁻¹ s          ＝ₛ⟨ seq⁻¹-sym-[↓] s ⟩
+                         ([ s ↓] ⁻¹) ◃∎   ＝↓⟨ ap (λ v → v ⁻¹) (＝ₛ-out e) ⟩
+                         ([ t ↓] ⁻¹) ◃∎   ＝ₛ⟨ sym-[↓]-seq⁻¹ t ⟩
+                         seq⁻¹ t ∎ₛ
+
+seq⁻¹-left-inverse : {x x' : X} (s : x ≡ x')
+                   → seq⁻¹ s ∙≡ s  ＝ₛ []
+seq⁻¹-left-inverse s = ＝ₛ-in (
```
Comment (martinescardo) [inline source/PathSequences/Inversion.lagda:63]: Same.

File: source/PathSequences/Inversion.lagda [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+                   [ (seq⁻¹ s ∙≡ s) ↓]   ＝⟨ ( [↓]-hom (seq⁻¹ s) s ) ⁻¹ ⟩
+                   [ seq⁻¹ s ↓] ∙ [ s ↓] ＝⟨ ap (_∙ [ s ↓]) (＝ₛ-out (seq⁻¹-sym-[↓] s)) ⟩
+                   [  s ↓] ⁻¹ ∙ [ s ↓]   ＝⟨ left-inverse ([ s ↓]) ⟩
+                   refl                   ∎ )
+
+seq⁻¹-right-inverse : {x x' : X} (s : x ≡ x')
+                    → s ∙≡ seq⁻¹ s ＝ₛ []
+seq⁻¹-right-inverse s = ＝ₛ-in ( [ s ∙≡ seq⁻¹ s ↓]     ＝⟨ [↓]-hom s (seq⁻¹ s) ⁻¹ ⟩
```
Comment (martinescardo) [inline source/PathSequences/Inversion.lagda:71]: Same.

File: source/PathSequences/Inversion.lagda [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+                                [ s ↓] ∙ [ seq⁻¹ s ↓] ＝⟨ ap ([ s ↓] ∙_) (＝ₛ-out (seq⁻¹-sym-[↓] s)) ⟩
+                                [ s ↓] ∙ [ s ↓] ⁻¹    ＝⟨ ( right-inverse [ s ↓] ) ⁻¹ ⟩
+                                 refl                  ∎ )
+
+\end{code}
+
+Alternative names
```
Comment (martinescardo) [inline source/PathSequences/Inversion.lagda:78]: Full stop.

File: source/PathSequences/Reasoning.lagda [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+  ＝-＝ₛ-equiv : (s t : x ≡ y) → (s ＝↓ t) ≃ (s ＝ₛ t)
+  ＝-＝ₛ-equiv s t = ＝ₛ-in , (＝ₛ-out , λ _ → refl) , (＝ₛ-out , λ _ → refl)
+
+\end{code}
+
+  TODO: there is a function
+
+  ＝ₛ-level : {s t : x ≡ y} {n : ℕ} → X is-of-level n → (s ＝ₛ t) is-of-level (n - 2)
+
+  which awaits implementation once I figure out how to deal with the global univalence in UF.HLlevels
```
Comment (martinescardo) [inline source/PathSequences/Reasoning.lagda:45]: Full stop. (Everywhere please.)

File: source/PathSequences/Reasoning.lagda [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+              → (m n : ℕ)
+              → {r : point-from-start m s ＝ point-from-start n (drop m s)}
+              → [ take n (drop m s) ↓] ＝ r
+              → take m s ∙≡ r ◃∙ drop n (drop m s) ＝ₛ u
+              → s ＝ₛ u
+  _＝↓⟨_&_&_⟩_ s {u} m n {r} p q = s ＝ₛ⟨ m &  n &  ＝ₛ-in p ⟩ q
+
+\end{code}
+
+  This closes
```
Comment (martinescardo) [inline source/PathSequences/Reasoning.lagda:214]: Full stop.

File: source/PathSequences/Reasoning.lagda [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+  
+  _∎ₛ : (s : x ≡ y) → s ＝ₛ s
+  _ ∎ₛ = ＝ₛ-in refl
+
+\end{code}
+
+Fixities
+
+\begin{code}
+  infixr 10 _＝ₛ⟨id⟩_
```
Comment (martinescardo) [inline source/PathSequences/Reasoning.lagda:226]: Blank line here please.

File: source/PathSequences/Rotations.lagda [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+
+post-rotate'-seq-in : {x y z : X} {r : x ≡ z} {q : y ≡ z} {p : x ≡ y}
+                    → r ＝ₛ p ∙≡ q
+                    → r ∙≡ (seq⁻¹ q) ＝ₛ p
+post-rotate'-seq-in {r = r} {[]} {p} e = r ∙≡ [] ＝ₛ⟨ []-∙≡-right-neutral-＝ₛ r ⟩
+                                         r       ＝ₛ⟨ e ⟩
+                                         p ∙≡ [] ＝ₛ⟨ []-∙≡-right-neutral-＝ₛ p ⟩
+                                         p       ∎ₛ
+post-rotate'-seq-in {r = r} {q ◃∙ s} {p} e =
+                    r ∙≡ (seq⁻¹ s ∙▹ (q ⁻¹))               ＝ₛ⟨ ＝ₛ-in (ap ≡-to-＝ (∙≡-assoc r (seq⁻¹ s) ((q ⁻¹) ◃∎) ⁻¹)) ⟩
```
Comment (martinescardo) [inline source/PathSequences/Rotations.lagda:126]: Indentation by 1 please.

**Review by martinescardo — CHANGES_REQUESTED** [review]
Very nice you ported this, which should be very useful! 

General stylistic comments: (1) We prefer indentation by 1 space everywhere (although some portions of `TypeTopology` currently violate this). (2) We prefer lines that are not longer than 80 characters (so that we can have at least to parallel code windows open). (3) We usually use one, rather than two lines to separate definitions. 

In any case, your code looks good. Thanks!

**Review by martinescardo — COMMENTED** [review]
One more thing: Please check your flags. Please remove `--auto-inline` (because it is in `typetopology.agda-lib`), however, please *keep* `--without-K` (for emphasis!) and `--safe`. Thanks!

**Comment by ettorealdrovandi** [comment]
I think I got all the changes done

**Pull request merged**